In [7]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [8]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib

from scipy.cluster.hierarchy import dendrogram, linkage, fcluster, inconsistent
from scipy.spatial.distance import cdist, pdist

import matplotlib.pyplot as plt
%matplotlib inline

algo = "HC"
base_path = "/content/drive/MyDrive/MIDA2/IntelligentMobilityProject/"
pictures_path = f"{base_path}Pictures/"
dataset_path = f"{base_path}Survey+dataset/"
code_path = f"{base_path}Code/"
results_path = f"{base_path}Code/Data/"
HC_base_path = f"{results_path}{algo}/"
numpy_file_type = ".npy"
image_file_type = ".html"

import sys
sys.path.append(code_path)
np.set_printoptions(precision=5, suppress=True)

# Merge cuts

In [9]:
np.set_printoptions(threshold=sys.maxsize)

In [10]:
metric = "VDM"
vars = 5
link = "complete"
n_clusters = 5
print(metric, vars, link)

VDM 5 complete


In [11]:
HC_path = f"{HC_base_path}{metric}{vars}/{algo}_{metric}_{link}{numpy_file_type}"
Z = {}
Z[metric] = np.load(HC_path)
cluster_labels = fcluster(Z[metric], n_clusters, criterion='maxclust')

In [12]:
non_zero_distance_merges = Z[metric][:,2] > 0
#print(Z[metric][non_zero_distance_merges])
print(Z[metric][non_zero_distance_merges].shape)

(3292, 4)


# Merge cuts

In [13]:
import postprocessing.merge_cuts
import importlib
importlib.reload(postprocessing.merge_cuts) 

<module 'postprocessing.merge_cuts' from '/content/drive/MyDrive/MIDA2/IntelligentMobilityProject/Code/postprocessing/merge_cuts.py'>

In [14]:
from postprocessing.merge_cuts import merge_multiple_cuts

In [15]:
k_values = [4, 6, 8]

merged_cuts = merge_multiple_cuts(Z[metric], k_values)
print(np.unique(merged_cuts, return_counts=True))

k1: [1 2 3 4]
k2: [5, 6]
k1: [ 1  2  3 10 11]
k2: [5, 6]
(array([ 1,  2, 10, 11, 17, 18], dtype=int32), array([1486,  514, 5734, 8969, 3279, 6623]))


# Plot silhouette

In [16]:
metric = "VDM"
vars = 6

In [17]:
ds_path = f'{dataset_path}4_DNA_{vars}values_normalized.csv'
df_DNA = pd.read_csv(ds_path)

In [18]:
import metrics.silhouette
import plots.silhouette
import importlib
importlib.reload(metrics.silhouette) 
importlib.reload(plots.silhouette)

<module 'plots.silhouette' from '/content/drive/MyDrive/MIDA2/IntelligentMobilityProject/Code/plots/silhouette.py'>

In [19]:
from metrics.silhouette import getSilhouette
from plots.silhouette import plotSilhouette
from plotly.subplots import make_subplots
from metrics.elbow import getElbows

In [20]:
link = "average"
postprocessing = False
is_postprocessing = "_fix" if postprocessing else ""
print(metric, vars, link, "postprocessing:", postprocessing)
range_min = 2
range_max = 100

VDM 6 average postprocessing: False


In [21]:
from plots.silhouette import plot_silhouette_merge_k_values

In [ ]:
vars_values = [5,6]
linkages = ["complete", "average", "weighted"]
postprocessing_values = [False, True]
for vars in vars_values:
  ds_path = f'{dataset_path}4_DNA_{vars}values_normalized.csv'
  df_DNA = pd.read_csv(ds_path)
  try:
    del distance_matrix
  except:
    distance_matrix = []
  distance_matrix = np.load(f"{results_path}distance_matrix_{metric}{vars}.npy")
  for link in linkages:
    WSS_path = f"{HC_base_path}{metric}{vars}/WSS/WSS_{metric}_{link}_range({range_min},{range_max}){numpy_file_type}"
    wss = np.load(WSS_path)
    k_values = getElbows(wss, range_min, add_first_k=True)
    HC_path = f"{HC_base_path}{metric}{vars}/{algo}_{metric}_{link}{numpy_file_type}"
    Z = {}
    Z[metric] = np.load(HC_path)
    for postprocessing in postprocessing_values:
      is_postprocessing = "_fix" if postprocessing else ""
      save_path = f"{pictures_path}HC/Silhouette/{metric}{vars}/MergedCuts/MergedCuts_{metric}_{link}_{str(k_values)}{is_postprocessing}.html"
      plot_silhouette_merge_k_values(df_DNA, distance_matrix, Z[metric], metric, link ,k_values, postprocessing, save_path=save_path)

/content/drive/MyDrive/MIDA2/IntelligentMobilityProject/Code/metrics/elbow.py:18: RuntimeWarning:

divide by zero encountered in double_scalars



For n_clusters = 2 The average silhouette_score is : 0.6858174526302973
k1: [1 2]
k2: [3, 4]
For n_clusters = 4 The average silhouette_score is : 0.30054295644899076
k1: [1 2]
k2: [3, 4]
k1: [1 2 6 7]
k2: [5, 6]
For n_clusters = 5 The average silhouette_score is : 0.26685395685491603
k1: [1 2]
k2: [3, 4]
k1: [1 2 6 7]
k2: [5, 6]
k1: [ 1  2  6 13 14]
k2: [5, 6]
For n_clusters = 6 The average silhouette_score is : 0.257384567501542
k1: [1 2]
k2: [3, 4]
k1: [1 2 6 7]
k2: [5, 6]
k1: [ 1  2  6 13 14]
k2: [5, 6]
k1: [ 1  2 13 14 20 21]
k2: [13, 14]
For n_clusters = 7 The average silhouette_score is : 0.20636338133930035
k1: [1 2]
k2: [3, 4]
k1: [1 2 6 7]
k2: [5, 6]
k1: [ 1  2  6 13 14]
k2: [5, 6]
k1: [ 1  2 13 14 20 21]
k2: [13, 14]
k1: [ 1  2 14 20 21 35 36]
k2: [10, 11]
For n_clusters = 8 The average silhouette_score is : 0.1888135119288355
k1: [1 2]
k2: [3, 4]
k1: [1 2 6 7]
k2: [5, 6]
k1: [ 1  2  6 13 14]
k2: [5, 6]
k1: [ 1  2 13 14 20 21]
k2: [13, 14]
k1: [ 1  2 14 20 21 35 36]
k2: [10, 

/content/drive/MyDrive/MIDA2/IntelligentMobilityProject/Code/metrics/elbow.py:18: RuntimeWarning:

divide by zero encountered in double_scalars



For n_clusters = 2 The average silhouette_score is : 0.6858174526302973
k1: [1 2]
k2: [2, 3]
For n_clusters = 3 The average silhouette_score is : 0.4407280676985771
k1: [1 2]
k2: [2, 3]
k1: [1 5 6]
k2: [4, 5]
For n_clusters = 4 The average silhouette_score is : 0.378107366107184
k1: [1 2]
k2: [2, 3]
k1: [1 5 6]
k2: [4, 5]
k1: [ 1  5 11 12]
k2: [6, 7]
For n_clusters = 5 The average silhouette_score is : 0.3167876358294557
k1: [1 2]
k2: [2, 3]
k1: [1 5 6]
k2: [4, 5]
k1: [ 1  5 11 12]
k2: [6, 7]
k1: [ 1  5 11 19 20]
k2: [8, 9]
For n_clusters = 6 The average silhouette_score is : 0.3125951667044843
k1: [1 2]
k2: [2, 3]
k1: [1 5 6]
k2: [4, 5]
k1: [ 1  5 11 12]
k2: [6, 7]
k1: [ 1  5 11 19 20]
k2: [8, 9]
k1: [ 1  5 11 19 29 30]
k2: [8, 9]
For n_clusters = 7 The average silhouette_score is : 0.28054958616496106
k1: [1 2]
k2: [2, 3]
k1: [1 5 6]
k2: [4, 5]
k1: [ 1  5 11 12]
k2: [6, 7]
k1: [ 1  5 11 19 20]
k2: [8, 9]
k1: [ 1  5 11 19 29 30]
k2: [8, 9]
k1: [ 1  5 11 29 30 39 40]
k2: [13, 14]
For n

/content/drive/MyDrive/MIDA2/IntelligentMobilityProject/Code/metrics/elbow.py:18: RuntimeWarning:

divide by zero encountered in double_scalars



For n_clusters = 2 The average silhouette_score is : 0.6858174526302973
k1: [1 2]
k2: [3, 4]
For n_clusters = 4 The average silhouette_score is : 0.378107366107184
k1: [1 2]
k2: [3, 4]
k1: [1 2 6 7]
k2: [3, 4]
For n_clusters = 5 The average silhouette_score is : 0.2777610664782623
k1: [1 2]
k2: [3, 4]
k1: [1 2 6 7]
k2: [3, 4]
k1: [ 1  6  7 11 12]
k2: [5, 6]
For n_clusters = 6 The average silhouette_score is : 0.2630477283540716
k1: [1 2]
k2: [3, 4]
k1: [1 2 6 7]
k2: [3, 4]
k1: [ 1  6  7 11 12]
k2: [5, 6]
k1: [ 1  6  7 11 18 19]
k2: [5, 6]
For n_clusters = 7 The average silhouette_score is : 0.24882116237562607
k1: [1 2]
k2: [3, 4]
k1: [1 2 6 7]
k2: [3, 4]
k1: [ 1  6  7 11 12]
k2: [5, 6]
k1: [ 1  6  7 11 18 19]
k2: [5, 6]
k1: [ 1  6  7 18 19 25 26]
k2: [10, 11]
For n_clusters = 9 The average silhouette_score is : 0.24147764990406032
k1: [1 2]
k2: [3, 4]
k1: [1 2 6 7]
k2: [3, 4]
k1: [ 1  6  7 11 12]
k2: [5, 6]
k1: [ 1  6  7 11 18 19]
k2: [5, 6]
k1: [ 1  6  7 18 19 25 26]
k2: [10, 11]
k1:

# Good results


## VDM5

|Link|Cuts|(No postprocessing) Sil Avg|(Postprocessing) Sil Avg|Outliers|
|------|-----|-----|-----|----|
|average|[2,3,5,7,9]|  0.31| 0.36 |131|
|average|[2,3,5,7,9,12]|  0.28| 0.33 |639|
|complete|[2,4]|  0.30| 0.34 |355|
|complete|[2,4,6]|  0.26| 0.33 |685|
|complete|[2,4,6,8]|  0.26| 0.3 |151|
|single|[2,4,6,33,49]|  0.367| 0.376 |600|
|weighted|[2,4,6,8]|  0.26| 0.33 |1039|
|weighted|[2,4,6,8,11]|  0.24| 0.33 |1480|

## VDM6

|Link|Cuts|(No postprocessing) Sil Avg|(Postprocessing) Sil Avg|Outliers|
|------|-----|-----|-----|----|
|average|[2,3,5,7,9]|  0.31| 0.35 |160|
|complete|[2,3,5]|  0.24| 0.30 |211|
|weighted|[2,4,6,8,11,13]|  0.25| 0.33 |532|